### PR TITLE
Adds ability to add parameter coercions in the IR.

### LINKED ISSUE #808: More parameter coercions should be in IR
Right() function truncates its 2nd arg 

https://github.com/microsoft/Power-Fx/blob/d01d9b7baf81c7a64fc356329f83a38c55a497a6/src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Right.txt#L6

Expected: "World". actual: "oWorld"

Currently, this is captured in the function's implementation - so each backend has to get it right.  
Instead, this should be captured in the IR. 

Ie, instead of an IR like:
Call(Right,  arg0: "HelloWorld", arg1: 5.8)
and then the impl of Right knows arg1 should be truncated.  (and some impls just get it wrong)

It should be an IR like:
Call(Right,  arg0: "HelloWorld", Truncate(arg1: 5.8))

_And then the impl of Right doesn't need to worry about this corner case._  So all backends will get it right. 
(TBD if Truncate is an existing operator or if we need to add a new one )

**Places where we can leverage this:**
- string functions like Right() that convert numeric argument to an interger. 
- Blank() to 0 
- Blank() to empty string .

This will require some sort of hook on the TexlFunction for translating the call node that the IR translator can use.

### PR DESCRIPTION
**Pull request by jas-valgotar**
- Fixes #808 
- Now, one can add param coercions in IR phase for a call node by overriding `CreateIRCallNode` method of TexlFunction. So, this brings an advantage that other backends would not need to worry about taking care of things like converting `Blank()` arguments to zero or empty string or similar preprocessing in implementation.
- Updated `Left()`, `Right()`, `Concatenate()`, and `Abs()` function to leverage this.

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs [hunk 1]
```diff
@@ -1445,5 +1453,57 @@ internal FunctionInfo Info(string locale)
...
             };
         }
+
+        /// <summary>
+        /// This can be overridden so function can add preprocessing capabilities at IR phase.
+        /// Consider using Helper function such as <see cref="BlankToZeroIRCallNode(IntermediateNode)"/>, 
+        /// <see cref="BlankToEmptyStringIRCallNode(IntermediateNode)"/>,
+        /// <see cref="NumberTruncateIRCallNode(IntermediateNode)"/> to add preprocessing when overriding.
+        /// </summary>
+        internal virtual IRCallNode CreateIRCallNode(IRContext context, List<IntermediateNode> args, IR.Symbols.ScopeSymbol scope)
```
[the remark this answers is not in the document]
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs:1463]: I agree that is potentially dangerous, but the challenge is that we want the flexibility to rewrite an expression for call node so we can use existing functions in combinations to wrap call node like `If(IsBlank(xCallNode), 0, xCallNode)`. Can you elaborate on the approach you are proposing?

[the remark this answers is not in the document]
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs:1463]: We looked at something more declarative, like returning an array of UnaryOpKind to apply to the parameters. 
But the unary ops don't match what we needed. 
But maybe introducing a private enum that maps to Blank20, Truncate, etc would do it.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs [hunk 2]
```diff
@@ -316,6 +321,57 @@ public override IntermediateNode Visit(TexlCallNode node, IRTranslatorContext co
...
+                switch (preProcessor)
+                {
+                    case IRPreProcessor.BlankToZero:
+                        return BlankToZeroIRCallNode(argIR);
+                    case IRPreProcessor.BlankToEmptyString: 
+                        return BlankToEmptyStringIRCallNode(argIR);
+                    case IRPreProcessor.NumberTruncate:
+                        return NumberTruncateIRCallNode(argIR);
+                    default:
+                        return argIR;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs:335]: nit: 
safest is to handle none explicitly and then for "default" to throw unhandled exception.

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs]: added an enum, so now TexlFunction doesn't need to return CallNode.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/Syntax/IRPreProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.PowerFx.Syntax
+{
+    internal enum IRPreProcessor
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Syntax/IRPreProcessor.cs:10]: >IRPreProcessor [](http://example.com/codeflow?start=18&length=14)

We should add a comment for this. 
Xml comments can also refer to specific function, so you can map the

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request closed without merging**